[dbt tutorial] update upstream assets section of dbt tutorial (#15556)
dagster-io/dagster: docs/content/integrations/dbt/using-dbt-with-dagster/upstream-assets.mdx

---

> title: "Using Dagster with dbt, part 3: Add assets upstream of your dbt models"

Reviewer: "Define" over "add"?
```suggestion
title: "Using Dagster with dbt, part 3: Define assets upstream of your dbt models"
```

Author: 👍

By this point, you've [set up a dbt project](/integrations/dbt/using-dbt-with-dagster/set-up-dbt-project) and [loaded dbt models into Dagster as assets](/integrations/dbt/using-dbt-with-dagster/load-dbt-models).

However, the tables at the root of the pipeline are static: they're [dbt seeds](https://docs.getdbt.com/docs/build/seeds), CSVs that are hardcoded into the dbt project. In a more realistic data pipeline, these tables would typically be ingested from some external data source, for example by using a tool like Airbyte or Fivetran, or by Python code.

> These ingestion steps in the pipeline often don't make sense to define inside dbt, but they often still do make sense to define as Dagster software-defined assets. You can think of a Dagster software-defined asset as a more general version of a dbt model. A dbt model is one kind of asset, but another kind is one that's defined in Python, using Dagster's Python API.

Author: I want to make sure that it's clear that "defining with these as an SDA" isn't communicated as mutually exclusive with "defining it inside dbt".  I.e. an asset defined in dbt is an SDA.

Like: "I don't live in Tulsa anymore, but I _still_ live in Oklahoma"

dbt_project_dir = Path(__file__).parent.joinpath("..", "..")
duckdb_database_path = dbt_project_dir.joinpath("tutorial.duckdb")

@asset
def raw_customers(context) -> None:
    data = pd.read_csv("https://docs.dagster.io/assets/customers.csv")
    connection = duckdb.connect(duckdb_database_path)
    connection.execute("create schema if not exists jaffle_shop")
    connection.execute("create or replace table jaffle_shop.raw_customers as select * from data")

    # Log some metadata about the table we just wrote. It will show up in the UI.
    context.add_output_metadata({"num_rows": data.shape[0]})

dbt = DbtCliResource(project_dir=os.fspath(dbt_project_dir))
```

Let's review the changes we made:

1. At the top, we added imports for `pandas` and `duckdb`, which we use to for fetching data into a `DataFrame` and writing it to DuckDB.

2. We added a `duckdb_database_path` variable, which holds the location of our DuckDB database. Remember that DuckDB databases are just regular files on the local filesystem. The path is the same path that we used when we configured our profiles.yml file. This variable is used in the implementations of the `raw_customers` asset.

> 3. We added a definition for the `raw_customers` table. The `raw_customers` function with the <PyObject object="asset" decorator /> decorator above it is the function that defines the `raw_customers` table. The implementation inside the function fetches data from the internet and writes it to a table in our DuckDB database. Similar to how running a dbt model executes a select statement, materializing this asset will execute this Python code.

Reviewer 2: The first two sentences of this paragraph tripped me up. After reading the first one, I thought it was going to talk about some of the SQL statements in the asset function as those technically 'create' the table.

I get what this is saying after reading it again, but perhaps combining these two sentences could reduce their redundancy and make the meaning more clear?

Finally, let's update the `assets` argument of our `Definitions` object, at the bottom of the file, to include the new asset we just defined:

```python file=integrations/dbt/tutorial/upstream_assets/definitions.py startafter=start_defs endbefore=end_defs
defs = Definitions(
    assets=[raw_customers, jaffle_shop_dbt_assets],
    schedules=schedules,
    resources={
        "dbt": dbt,
    },
)
```

---

> ## Step 2: In the dbt project, replace a seed with a source

Reviewer 2: The code blocks are rendering funny in this whole section, likely because they're not indented under the line items correctly. Could you address this? LMK if you need help, since Markdown can be finicky about this

1. Because we're replacing it with a Dagster asset, we no longer need the dbt seed for `raw_customers`, so we can delete it:

```shell
cd ..
rm seeds/raw_customers.csv
```

> 2. Instead, we want to tell dbt that `raw_customers` is a table that is defined outside of the dbt project. We can do that by defining it inside a [dbt source](https://docs.getdbt.com/docs/build/sources). Create a file called `sources.yml` inside the `models/` directory, and put this inside it:

Reviewer 2: ```suggestion
2. Instead, we want to tell dbt that `raw_customers` is a table that is defined outside of the dbt project. We can do that by defining it inside a [dbt source](https://docs.getdbt.com/docs/build/sources). 

   Create a file called `sources.yml` inside the `models/` directory, and put this inside it:
```

Suggesting we add a line break here to make the instruction stand out more

    tables:
      - name: raw_customers
        meta:
          dagster:
            asset_key: raw_customers
```

This is a standard dbt source definitions, with one addition: it includes metadata, under the `meta` property, that specifies the Dagster asset that it corresponds to. When Dagster reads the contents of the dbt project, it reads this metadata and draws the correspondence. For any dbt model that depends on this dbt source, Dagster then knows that the Dagster asset corresponding to the dbt model should depend on the Dagster asset corresponding to the source.

3. Then, update the model that depends on the `raw_customers` seed to instead depend on the source. Replace the contents of `model/staging/stg_customers.sql` with this:

```sql
with source as (

    {#-
    Use source instead of seed:
    #}
    select * from {{ source('jaffle_shop', 'raw_customers') }}

),

renamed as (

    select
        id as customer_id,
        first_name,
        last_name

    from source

)

select * from renamed
```

4. Now let's reload our data pipeline in the UI to see what we've built:

```shell
cd jaffle_dagster/

DAGSTER_DBT_PARSE_PROJECT_ON_LOAD=1 dagster dev
```

> ## Step 3: Materialize the assets using the UI

Reviewer 2: ```suggestion
---

## Step 3: Materialize the assets using the UI
```

Adding a missing rule

Now that you've created assets and resources, it's time to materialize the assets! Materializing an asset runs the op it contains and saves the results to persistent storage. In this tutorial, we're saving asset outputs to DuckDB.

1. In the UI, click the **Reload definitions** button. This ensures that the UI picks up the changes you made in the previous steps.

>    At this point, the `raw_customers` asset should display above `stg_customers` as upstream dependencies:

Reviewer 2: > should display above `stg_customers` as upstream dependencies:

Should this be `should display above stg_customers as an upstream dependency?`

   <Image

>    alt="Asset graph in the Dagster UI, showing dbt models and unmaterialized assets"

Author: Screenshots have not been updated, because of the "stumped" issue I mentioned in the PR description.

By this point, you've [set up a dbt project](/integrations/dbt/using-dbt-with-dagster/set-up-dbt-project) and [loaded dbt models into Dagster as assets](/integrations/dbt/using-dbt-with-dagster/load-dbt-models).

However, the tables at the root of the pipeline are static: they're [dbt seeds](https://docs.getdbt.com/docs/build/seeds), CSVs that are hardcoded into the dbt project. In a more realistic data pipeline, these tables would typically be ingested from some external data source, for example by using a tool like Airbyte or Fivetran, or by Python code.

> These ingestion steps in the pipeline often don't make sense to define inside dbt, but they often still do make sense to define as Dagster software-defined assets. You can think of a Dagster software-defined asset as a more general version of a dbt model. A dbt model is one kind of asset, but another kind is one that's defined in Python, using Dagster's Python API.

Reviewer: "often" gives indicates uncertainty here imo. can we remove it?

Finally, let's update the `assets` argument of our `Definitions` object, at the bottom of the file, to include the new asset we just defined:

```python file=integrations/dbt/tutorial/upstream_assets/definitions.py startafter=start_defs endbefore=end_defs
defs = Definitions(
    assets=[raw_customers, jaffle_shop_dbt_assets],
    schedules=schedules,
    resources={
        "dbt": dbt,
    },
)
```

---

## Step 3: In the dbt project, replace a seed with a source

1. Because we're replacing it with a Dagster asset, we no longer need the dbt seed for `raw_customers`, so we can delete it:

   ```shell
   cd ..
   rm seeds/raw_customers.csv
   ```

2. Instead, we want to tell dbt that `raw_customers` is a table that is defined outside of the dbt project. We can do that by defining it inside a [dbt source](https://docs.getdbt.com/docs/build/sources).

   Create a file called `sources.yml` inside the `models/` directory, and put this inside it:

   ```yaml
   version: 2

   sources:
     - name: jaffle_shop
       tables:
         - name: raw_customers
           meta:
             dagster:
               asset_key: raw_customers
   ```

> This is a standard dbt source definitions, with one addition: it includes metadata, under the `meta` property, that specifies the Dagster asset that it corresponds to. When Dagster reads the contents of the dbt project, it reads this metadata and draws the correspondence. For any dbt model that depends on this dbt source, Dagster then knows that the Dagster asset corresponding to the dbt model should depend on the Dagster asset corresponding to the source.

Reviewer: ```suggestion
This is a standard dbt source definition, with one addition: it includes metadata, under the `meta` property, that specifies the Dagster asset that it corresponds to. When Dagster reads the contents of the dbt project, it reads this metadata and infers the correspondence. For any dbt model that depends on this dbt source, Dagster then knows that the Dagster asset corresponding to the dbt model should depend on the Dagster asset corresponding to the source.
```
